src/process_utility.c

```diff
@@ -472,6 +474,225 @@ process_drop_trigger_start(ProcessUtilityArgs *args, DropStmt *stmt)
+				ereport(NOTICE, errmsg("drop cascades to job %d", job_id));
+				ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+				ts_catalog_restore_user(&sec_ctx);
+			}
+		}
+	}
+	return DDL_CONTINUE;
+}
+
+/*
+ * Start of dropping a procedure.
+ *
+ * We can abort the drop here by throwing an error.
+ */
+static void
+process_drop_procedure_start(DropStmt *stmt)
+{
+	ScanIterator iterator =
+		ts_scan_iterator_create(BGW_JOB, RowExclusiveLock, CurrentMemoryContext);
+	ts_scanner_foreach(&iterator)
+	{
+		ListCell *cell;
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		bool schema_isnull, name_isnull, job_id_isnull;
+		Name proc_schema =
+			DatumGetName(slot_getattr(ti->slot, Anum_bgw_job_proc_schema, &schema_isnull));
+		Name proc_name = DatumGetName(slot_getattr(ti->slot, Anum_bgw_job_proc_name, &name_isnull));
+		int32 job_id = DatumGetInt32(slot_getattr(ti->slot, Anum_bgw_job_id, &job_id_isnull));
+		Ensure(!job_id_isnull, "corrupt job entry: job id was null");
+		Ensure(!schema_isnull, "corrupt job entry: schema for job %d was null", job_id);
+		Ensure(!name_isnull, "corrupt job entry: name for job %d was null", job_id);
+		TS_DEBUG_LOG("looking at job %d with %s.%s",
+					 job_id,
+					 NameStr(*proc_schema),
+					 NameStr(*proc_name));
+		foreach (cell, stmt->objects)
+		{
+			ObjectWithArgs *object = castNode(ObjectWithArgs, lfirst(cell));
+			RangeVar *rel = makeRangeVarFromNameList(object->objname);
+			if (namestrcmp(proc_schema, rel->schemaname) == 0 &&
+				namestrcmp(proc_name, rel->relname) == 0)
+			{
+				Assert(stmt->removeType == OBJECT_PROCEDURE || stmt->removeType == OBJECT_FUNCTION);
+				if (stmt->behavior == DROP_RESTRICT)
+				{
+					ereport(ERROR,
+							errcode(ERRCODE_DEPENDENT_OBJECTS_STILL_EXIST),
+							errmsg("cannot drop %s because background job %d depends on it",
+								   NameListToString(object->objname),
+								   job_id),
+							errhint("Use delete_job() to drop the job first."));
+				}
+				else
+				{
+					CatalogSecurityContext sec_ctx;
+					ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+					ereport(NOTICE, errmsg("drop cascades to job %d", job_id));
+					ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+					ts_catalog_restore_user(&sec_ctx);
+				}
+			}
+		}
+	}
+}
+
+static void
+replace_attr_if_changed(AttrNumber attno, const char *newvalue, Name name_buf, Datum *values,
+						bool *replace)
+{
+	if (newvalue)
+	{
+		const Name orig_value = DatumGetName(values[AttrNumberGetAttrOffset(attno)]);
+		if (namestrcmp(orig_value, newvalue) != 0)
+		{
+			namestrcpy(name_buf, newvalue);
+			values[AttrNumberGetAttrOffset(attno)] = NameGetDatum(name_buf);
+			replace[AttrNumberGetAttrOffset(attno)] = true;
+		}
+	}
+}
+
+/*
+ * Update the schema or name of a procedure in the jobs tuple.
+ */
+static void
+ts_bgw_job_update_proc(Relation rel, HeapTuple tuple, TupleDesc tupledesc, const char *newschema,
```

Comment:
The following functions seem out-of-place here. Shouldn't they be in `bgw/`?

Reply:
Also, usually we use `ts_` prefix only for "public" functions.

Reply:
This is the pattern followed for the other functions, so just tagging along here. See, for example `process_rename_schema`.